physlr/physlr.py

```diff
@@ -2236,24 +2237,216 @@ def physlr_bed_to_path(self):
+            temp = int_columns[3]
+            int_columns[3] = int_columns[4]
+            int_columns[4] = temp
+            pairs[(columns[1], columns[0])] = int_columns[2:6]
+        arks_pair_file.close()
+        print("Read ARCS pair", file=sys.stderr)
+        return pairs
+
+    @staticmethod
+    def read_dist_est(filename, dist_type):
+        """
+        Read ARCS distance estimate. Return a dictionary of pairs to distance estimates.
+        """
+        dist = {}
+        dist_est_file = open(filename, "r")
+
+        dist_type_to_idx = {"min":2, "avg":3, "max":4}
+
+        if dist_type not in dist_type_to_idx:
+            print("invalid --dist-type parameters", file=sys.stderr)
+            sys.exit(1)
+        idx = dist_type_to_idx[dist_type]
+        for line in dist_est_file:
+            columns = line.rstrip().split("\t")
+            if columns[0] == "contig1":
+                continue
+
+            contig1 = columns[0].rstrip("-+")
+            contig2 = columns[1].rstrip("-+")
+            dist[(contig1, contig2)] = int(columns[idx])
+            dist[(contig2, contig1)] = int(columns[idx])
+
+        dist_est_file.close()
+        print("Read Distance Estimates", file=sys.stderr)
+        return dist
+
+    @staticmethod
+    def normal_estimation(x, probability, n):
+        """
+        Compute binomial estimation
+        """
+        mean_val = n * probability
+        std_dev = float(math.sqrt(mean_val * (1 - probability)))
+        return 0.5 * (1 + math.erf((x - mean_val) / (std_dev * math.sqrt(2))))
+
+    @staticmethod
+    def is_link_significant(ori_list):
+        """
+        Check if barcode link is strong enough
+        """
+        max_idx = ori_list.index(max(ori_list))
+        ori_list.sort()
+        max_val = ori_list[-1]
+        sum_with_second_max = ori_list[-1] + ori_list[-2]
+        normal_cdf = Physlr.normal_estimation(max_val, 0.5, sum_with_second_max)
+        if 1 - normal_cdf < 0.05:
+            return max_idx
+        return -1
+
+    @staticmethod
+    def orient_part_of_path_backward(pairs, path, unoriented, curr_pos, name):
+        """
+        Orient small part of path based on ARCS scaffold pairing information going backwards
+        """
+        idxtojoin = {0:"-+", 1:"--", 2:"++", 3:"+-"}
+        while unoriented:
+            prev_pos = curr_pos - 1
+            pair = (path[prev_pos][:-1], name[:-1])
+            if pair in pairs:
+                join_ori = pairs[pair]
+                max_idx = Physlr.is_link_significant(join_ori)
+                if max_idx != -1:
+                    curr_ori = idxtojoin[max_idx][1]
+                    if curr_ori == name[-1]:
+                        prev_ori = idxtojoin[max_idx][0]
+                        path[prev_pos] = path[prev_pos][0:-1] + prev_ori
+                        del unoriented[-1]
+                        curr_pos = prev_pos
+                        name = path[prev_pos]
+                    else:
```

Comment:
Hmm lots of repeated `unoriented.clear()`
What if you have a flag which indicates if the piece was oriented or not, and have an if statement checking the flag right before being done the for-loop code?

Reply:
Wouldn't the number of lines still be the more? You'd be replacing each `clear()` with a flag, and two more lines to check the flag and `clear()`

Reply:
Hmm no you would only change the flag when you do the `.pop()`, and then would only have to check the flag one time when you are out of that if statement. Don't you only not clear if you enter the `if curr_ori == name[-1]:` statement?

Reply:
done